Implement simple audio transcription

Rafer45 opened the pull request.

apascal07 commented on storage-transcribe-audio/functions/src/transcribe-audio.ts:
```diff
@@ -0,0 +1,52 @@
+  client: SpeechClient, object: ObjectMetadata): Promise<void> {
+
+  const request = {
+    config: {
```
Are these user-configurable? Should they be params?

Rafer45 replied:
Some of this will depend on the sample rate of the specific file; the encoding can always be LINEAR16 because best practices ask us to transcode it (https://cloud.google.com/speech-to-text/docs/best-practices).

The language code should be configurable, though; yeah. Let's talk more about this off the issue

apascal07 replied:
Let's make LINEAR16 and 48000 constants near the top of the file.

Rafer45 replied:
LINEAR16 can be a constant at the top, but 48000 will come from the ffprobe call

apascal07 commented on storage-transcribe-audio/functions/src/transcribe-audio.ts:
```diff
@@ -0,0 +1,138 @@
+  };
+}
+
+// For files whose length we don't know, or whose length is 60 secs or more.
```
How does the implementation differ if the length is >60s vs <60s?

Rafer45 replied:
For <60s, we can call a different function that runs faster -- see ["transcribe short audio files"](https://cloud.google.com/speech-to-text/docs/sync-recognize) vs ["transcribe long audio files"](https://cloud.google.com/speech-to-text/docs/async-recognize)

Rafer45 replied:
After discussion, we've agreed to use <60s for all use cases, to use the API's upload capabilities.

huangjeff5 replied:
Huh does this mean we can't transcribe any files >60s?

huangjeff5 replied:
Wondering what the tradeoff was and why we decided with this.

Rafer45 replied:
Ah! Typo! We've agreed to use >60s for all use cases, to use the upload capabilities of the API 😅

huangjeff5 replied:
oh gotcha, cool.

huangjeff5 replied:
I'm still not sure why we can't use the sync endpoint for <60s, I'm assuming it's because we can't know if the file is <60s unless we download it? If that's the case, we can tell developers they have to attach a metadata field to objects uploaded containing the duration (otherwise, it's assumed that its >60s). But up to you if you feel the complexity is worth it.

Rafer45 replied:
We aren't using the sync endpoint for <60s because we're using a feature that's only available for the >60s endpoint. The feature we're using supports uploading results directly from the API to a Cloud Storage bucket, without piping the results through the extension; Alex recommended we use it because it's more resistant to failures. That feature isn't available for <60s.

Downloading the file isn't an issue because we download the file to transcode it for ease of use, so that the user can just upload any kind of audio file.

apascal07 commented on storage-transcribe-audio/functions/src/transcribe-audio.ts:
```diff
@@ -0,0 +1,138 @@
+  // TODO(reao): Handle video files
+}
+
+type TranscodeAudioResult = TranscodeAudioSuccess | TranscodeAudioFailure;
```
Let's maybe put all of these types and interfaces near the top.

Rafer45 replied:
SGTM; I'll let you know when I've done that

Rafer45 replied:
Moved the interfaces to their own file.

reao-creator committed:
Add params and config

reao-creator committed:
Add types

reao-creator committed:
Add audio transcoding and transcription

huangjeff5 commented on storage-transcribe-audio/extension.yaml:
```diff
@@ -1 +1,135 @@
+  - role: storage.admin
+    reason: Allows the extension to write to your Cloud Storage.
+  - role: datastore.user
+    reason: Allows the extension to write to your Firestore Database instance.
```
add events
transcription completed / failed

huangjeff5 replied:
If it failed it's also worth emitting the error in the event probably

Rafer45 replied:
I have a question about this, let's chat about it 1:1

reao-creator committed:
Emit events on transcription

reao-creator committed:
Use 'transcription' consistently in extension

reao-creator committed:
Emit event on error, log errors only on index.ts

Rafer45 commented on storage-transcribe-audio/functions/src/index.ts:
```diff
@@ -1,9 +1,141 @@
+        uploadResponse: [file /* metadata */],
+      } = transcodeResult;
+
+      const transcriptionResult = await transcribeAndUpload({
```
The warnings from the transcoder are forgotten here; I have a commit ready to fix that so that they are kept.

Rafer45 replied:
Done.

apascal07 commented on storage-transcribe-audio/functions/src/index.ts:
```diff
@@ -1,9 +1,141 @@
+      if (transcodeResult.state === "failure") {
+        logs.transcodingFailed(transcodeResult);
+        eventChannel &&
+          (await publishFailureEvent(eventChannel, transcodeResult));
```
Is this a common pattern? I find this less readable than having an if statement.

Rafer45 replied:
I copied the pattern from resize images, but I agree it's less readable than an if statement, so I'll change it to your suggestion.

Fixed.

apascal07 commented on storage-transcribe-audio/functions/src/transcribe-audio.ts:
```diff
@@ -0,0 +1,212 @@
+  logs.receivedLongRunningRecognizeResponse(response);
+  const taggedTranscription = response.results?.map(
+    (result) =>
+      [result?.channelTag, result?.alternatives?.[0].transcript] as const
```
Is the optional chaining on `result` here needed? If we're iterating over items, they shouldn't be null/undefined? This is hard to tell what it does, especially the [0] index, maybe breaking it out into variables would be more readable/easier to tell if the logic is flawed.

Rafer45 replied:
As far as the type system is concerned, these could be null/undefined; I don't know if the type system is lying to me, so I figured I'd chain on optionals, then use a type guard to ensure there's no nulls in the final output.

I could make a helper method that gets the first alternative transcript to make that second part more readable, with a comment explaining why there's a [0] index. I will try that and you can tell me what you think of the result.

Rafer45 replied:
Done, please let me know what you think.

Rafer45 replied:
Resolving, can clean up in follow-up.

apascal07 commented on storage-transcribe-audio/functions/src/transcribe-audio.ts:
```diff
@@ -0,0 +1,212 @@
+    };
+  }
+
+  // I simplify the transcription compared to the one that's usually given
```
Is this a TODO/will be removed before checking in the final version?

Rafer45 replied:
Mainly a question for reviewers 😅

Rafer45 replied:
I.e. I was asking the reviewers a question within the comment. What do you think of the question in the comment? We can also discuss it in 1:1 chat.

apascal07 replied:
I actually don't think I understand the problem/solution here. Can you explain more in depth?

Rafer45 replied:
Yes.

I have two options: The extension can either (a) make the API call, get the transcription from it, and upload it itself, or (b) include an output path in the API call, so that the speech-to-text API uploads the file. In discussion, we agreed that going with (b) would likely be more reliable.

However, the output of the API looks something like this:
```json
{
  "results": [
    {
      "channelTag": 0,
      "alternatives": [
        {
          "confidence": 0.98267895,
          "transcript": "how old is the Brooklyn Bridge",
        }
      ]
    },
    {
      "channelTag": 1,
      "alternatives": [
        {
          "confidence": 0.98267895,
          "transcript": "where are my glasses",
        }
      ]
    }
  ]
}
```

I think that that's needlessly cumbersome for the user, since the extension doesn't allow for alternatives anyways; I think the user would prefer something like this:
```json
[
  ["how old is the Brooklyn Bridge"],
  ["where are my glasses"],
]
```

But we can't create a simpler output if we're relying on the API to output the results. So the decision is, basically, what's more worth it? The reliability of the upload, or the simplicity of the output?

Rafer45 replied:
Having said that, it might be worth it to just drop support for multiple channels, since the most common use case probably won't involve them. That discussion's a bit out of scope, though, and it's independent of the question re:simplification.

Let me address all your review comments and see if we can merge this.

Rafer45 replied:
Resolving, we can have further discussion in a subsequent PR.

apascal07 commented on storage-transcribe-audio/functions/src/transcribe-audio.ts:
```diff
@@ -0,0 +1,212 @@
+    };
+  }
+
+  const localOutputPath = localCopyPath + TRANSCODE_TARGET_FILE_EXTENSION;
```
Is this guaranteed to be a well formed path? i.e. with a slash between the two pieces

Rafer45 replied:
The local copy path is given by the caller, so there's no strong guarantees that _that_ is well formed, but the concatenation will be fine if localCopyPath is well formed, because `TRANSCODE_TARGET_FILE_EXTENSION` is just some file extension (it's `'.wav'`)

apascal07 commented on storage-transcribe-audio/functions/src/transcribe-audio.ts:
```diff
@@ -0,0 +1,212 @@
+  logs.debug("uploaded transcoded file");
+
+  return {
+    state: "success",
```
What if the upload fails? This function feels like it should be separated into 1) transcribe, and 2) upload.

Rafer45 replied:
I agree. I will split it and get back to you.

Rafer45 replied:
Done.

apascal07 commented on storage-transcribe-audio/functions/src/transcribe-audio.ts:
```diff
@@ -0,0 +1,212 @@
+}
+
+async function transcodeLocally({
+  inputPath,
```
Is there any reason this can't just be `function transcodeLocally(inputPath: string, outputPath: string): Promise<string>`?

Rafer45 replied:
I don't like having two arguments that are of the same type such that order matters, seems easy to introduce a bug. I prefer having a label for each arg since they're easily accidentally flipped.

apascal07 replied:
The method signature becomes unwieldy like in `transcodeToLinear16AndUpload` where some arguments are in an object and others aren't. Every library in the world has methods that have multiple arguments with the same type that can be accidentally flipped; it's the responsibility of the caller to look at the method signature and provide the correct inputs. After all, this doesn't actually stop someone from passing in the wrong data, they would just make a different mistake.

Rafer45 replied:
TS/JS best practices recommend avoidance of multiple arguments of the same type (DMed you a link since it's internal). To quote, "Strive not to force call site readers to go to the method definition to understand the call". It recommends option bags, which look like what I've got, I think. The ease of making objects in typescript & the availability of destructuring makes the tradeoff more worth it than it would be in, like, Java or C++.

Maybe a better solution would be for me to include `bucket` in the option bag; that might be more consistent. What do you think?

If you do feel strongly about this, I'm willing to go with your suggestion of two string arguments, especially since the function is private.

apascal07 replied:
Ah, I see. If it's idiomatic to TS then that's okay. That's what I wanted to confirm. In any case, I think once you have a bag, you should put all the arguments in there, not just some like in the case of `transcodeToLinear16AndUpload`. I think this is consistent with our other OSS work.

Rafer45 replied:
SGTM, I will do that and get back to you.

Rafer45 replied:
Done.

apascal07 commented on storage-transcribe-audio/functions/src/types.ts:
```diff
@@ -0,0 +1,64 @@
+export type TranscribeAudioResult = TranscribeAudioSuccess | Failure;
+
+export interface TranscodeAudioSuccess {
+  state: "success";
```
Can we do an enum instead?

Rafer45 replied:
Yes.

Done.

apascal07 commented on storage-transcribe-audio/functions/src/index.ts:
```diff
@@ -1,9 +1,141 @@
+const eventChannel: Channel | null = process.env.EXT_SELECTED_EVENTS
+  ? getEventarc().channel(process.env.EXT_SELECTED_EVENTS, {
+      allowedEventTypes: process.env.EXT_SELECTED_EVENTS,
+    })
```
This doesn't look right, can you verify the arguments for this method? I think it just needs the second argument.

Rafer45 replied:
Ah, you're right, it was wrong. Does the updated version look better? I'm basing it off of image-resize.

apascal07 replied:
If `process.env.EVENTARC_CHANNEL` isn't set, should it use the default channel anyway?

Rafer45 replied:
I don't really know. Is there someone we could ask about this?

Rafer45 replied:
Also, this doesn't seem to be reflecting the latest version of the file; that's very confusing...

Rafer45 replied:
This is the practice followed by the Resize Images extension

Rafer45 replied:
Resolving, we can have further discussion in a subsequent PR.

apascal07 commented on storage-transcribe-audio/functions/src/transcribe-audio.ts:
```diff
@@ -0,0 +1,212 @@
+    },
+
+    outputConfig: {
+      gcsUri: `gs://${bucket.name}/${name}_transcription.txt`,
```
Let's maybe do these as constants at the top.

Rafer45 replied:
Done.

reao-creator committed:
Make eventChannel null check more readable

reao-creator committed:
Clarify & simplify tagged transcript logic

reao-creator committed:
Replace strings with enums

reao-creator committed:
Split transcrode step and upload step

apascal07 commented on storage-transcribe-audio/functions/src/transcribe-audio.ts:
```diff
@@ -61,14 +62,19 @@ export async function transcribeAndUpload({
+    return {
+      status: Status.FAILURE,
+      warnings,
+      type: FailureType.NULL_TRANSCRIPTION,
```
Is "null transcription" indicative of just any empty transcription? Like if I passed in an audio file that's just silent, what would be the output? Would it fail as is?

Rafer45 replied:
If the transcription is an empty string (as it would be if someone passed a silent audio file), a warning is emitted, but there's no error. A null transcription happens if the response doesn't have results, or if one of the results or channels is null, which I think would only be caused by a bug in the API (but still seems, well, possible).

apascal07 commented on storage-transcribe-audio/functions/src/transcribe-audio.ts:
```diff
@@ -61,14 +62,19 @@ export async function transcribeAndUpload({
+    };
+  }
+
+  const taggedTranscription: ([number, string]|null)[] = response.results.map(getTaggedTranscriptionOrNull)
```
This is not an ideal data type. Ideally it should be a map from channel tag to transcription. What does null signify here? An error or an empty transcription (e.g. silent audio)? I would refactor the method to return a map so you don't have to do the map here after returning the two piece array.

Rafer45 replied:
This is just an intermediate step. An array item will be null if either the channel or the transcript are null; there will never be nulls in the output because we check `!isNullFreeList` afterwards.

The final result is a map from channel tag to transcription arrays.

apascal07 replied:
I still think it's less readable than it needs to be because of the awkward types. You could have `getTaggedTranscriptions(response.results)` and inside just build the map or combine `separateByTags` from below and just return null if there are any nulls. It would require fewer in-between states. A list of tuples that are meant to be a map isn't a good type in my opinion, especially when you have the flexibility to return what you want.

Rafer45 replied:
I've changed it so that less of the logic is here. How does this look?

Rafer45 replied:
Resolving, we can have further discussion in a subsequent PR.

reao-creator committed:
Fix wrong eventarc params

reao-creator committed:
Delete bad tests config file

reao-creator committed:
Wrap failed uploads in a Failure

reao-creator committed:
Add comment to var with odd type

reao-creator committed:
Move strings to constants further up in expression tree

reao-creator committed:
Keep warnings from previous steps

reao-creator committed:
Move more unpacking logic to util.ts

reao-creator committed:
Remove output types param

Reducing scope of the extension so it can be finished

reao-creator committed:
Use outputCollection from config

reao-creator committed:
Remove buggy warning handler

This problem is a bit more complex than I initially thought and will
take a bit more design
